### ISSUE TITLE
Mutual-TLS Client Authentication

### ISSUE THREAD

**Issue by muhlemmer** [zitadel/zitadel#8597]
As an integrator I want to be able to perform Mutual-TLS client authentication as specified in [RFC8705, section 2](https://www.rfc-editor.org/rfc/rfc8705.html). This issue excludes "Certificate-Bound Access Tokens" from the same RFC.

### TLS termination

Many deployments of zitadel, including our cloud, terminate TLS in a reverse proxy.
I don't think this would be possible if we want mTLS support, unless we somehow forward the client's TLS credentials.

I don;t have a very deep knowledge about this parts of the TLS protocols. Perhaps @fforootd can shine a light on this subject?

Alternatively we push mTLS support to be handled by the proxy. In that case it's not a matter of the zitadel product, but more an operational problem.

### Certificate storage

mTLS is hard in a multi-instance setting. The TLS handshake happens when the HTTP connection is opened. By the time the request hits the business logic, we can only check what happened during the handshake through a [tls.ConnectionState](https://pkg.go.dev/crypto/tls#ConnectionState) in the HTTP request. As we share the HTTP server between OID/Oauth, gRPC, saml and others, a modification in the server config might impact all routes.

The go [http.Server](https://pkg.go.dev/net/http#Server) takes a [tls.Config](https://pkg.go.dev/crypto/tls#Config). In there we could set `ClientAuth = RequestClientCert`:

> // RequestClientCert indicates that a client certificate should be requested
	// during the handshake, but does not require that the client send any
	// certificates.

Once set, we will have the first challenge. The HTTP server will try to use `ClientCAs`, a CertPool which only allows setting certificates, no dynamic retrieval from storage. There is a fallback `VerifyPeerCertificate` function which may be used to lookup certificates from storage.

**Questions**

@livio-a, @adlerhurst as architects.

1. Do we want to store PKI root certificates on instance level?
2. Do we want to support and store self-signed certificates on application level?

1. Might be solved by an in-memory projection. Or even reduce directly into the `CertPool` and will have minimal performance impact.
2. Will require a DB query for the client during the TLS handshake, once a certificate is presented.

**Comment by fforootd** [comment]
I quickly refreshed my memory on mTLS (forwarding) 😁 

TL;DR I think we need to implement a configurable HTTP header for different proxies and the trust store to verify the chain.
Since we generally advise to run Zitadel behind a proxy I think it is important to support that scenario, otherwise we lose a lot of network protection capabilities. 

## Behind a "proxy"

There are multiple approaches that are "established". Most of them rely on the proxy layer to authenticate the connection and then forward the certificate and sometimes the chain as HTTP headers to the upstream system.

Keycloak for example is documented [here](https://www.keycloak.org/server/reverseproxy#_enabling_client_certificate_lookup) 

Another example is Cloudflare which is documented [here](https://developers.cloudflare.com/ssl/client-certificates/enable-mtls/#cloudflare-api)

Sad thing is that the encoding might be different with each proxy 😢 

A big question we need to ask here is if we need to verify the certificate against a trust chain again (which then needs to be available in Zitadel) or if we silently accept whatever comes from the proxy.
If we DO NOT verify the trust chain again we are at risk of an outside attacker sending the headers through the proxy (meaning they are not deleted at the proxy layer).
So I think we need to have trust verification at least optional 😁 

## Directly exposing

If we directly expose Zitadel we need a way of populating the trust store programmatically, which you outlined correctly above already.

**Comment by JayPe69** [comment]
Hello do you know if this feature will be implemented ?
Still under investigation ?

**Comment by muhlemmer** [comment]
> Hello do you know if this feature will be implemented ?

Perhaps, in the future if there is demand from community or enterprise costumers.

> Still under investigation ?

Given the complexities, laid out above, for which we do not have a proper solution yet, this remains under investigation.

**Comment by JayPe69** [comment]
@muhlemmer 

> Many deployments of zitadel, including our cloud, terminate TLS in a reverse proxy.

With Zitadel, you can configure Zitadel "web server" to enable TLS by default.

```
ZITADEL_EXTERNALPORT = "443"
ZITADEL_EXTERNALSECURE="true"
ZITADEL_TLS_ENABLED="true"
ZITADEL_TLS_KEYPATH="/certs/private_key.pem"
ZITADEL_TLS_CERTPATH="/certs/fullchain.pem"
```

So in that case, you have a full TLS chain.
The reverse proxy serve TLS for public endpoint, and internally the Zitadel webserver serve TLS too with an internal and trust CA.

To configure mTLS, we can ask the reverse proxy to manage it.
It means there will be mTLS from client to reverse proxy, and TLS from reverse proxy to Zitadel webserver.

This is not a full mTLS chain, but that's better than nothing
